### ISSUE TITLE
Ollama + Qwen3でLoraを作る環境を構築

### ISSUE THREAD

**Issue by szgk** [szgk/AITuber#4]
## 概要
WSL上のOllama + Qwen3で、AITuber用に学習したモデルを作りたい

## 方法
下記はChatGPTに聞いた手順なので間違ってる可能性あり

---

```markdown
# Qwen3-4B での LoRA 作成手順

このドキュメントでは、[Qwen3-4B](https://huggingface.co/Qwen/Qwen3-4B-Instruct) をベースに、LoRA（Low-Rank Adaptation）を使ってカスタムモデル（例：人格付き応答モデル）を学習・構築する方法を説明します。

---

## ✅ 準備環境

- Python 3.10+
- GPU（VRAM 8GB以上推奨）
- CUDA環境（Windows / WSL2 / Linux 対応）

### 必要なPythonパッケージ

```bash
pip install transformers datasets peft accelerate bitsandbytes
```

---

## ✅ 1. 学習データの準備（SFT形式）

LoRAは**命令応答形式の学習（Instruction Tuning）**に向いています。

### 学習データの例（JSONL形式）

```json
{"instruction": "こんにちはって言ってください", "input": "", "output": "こんにちは！"}
{"instruction": "あなたの名前は？", "input": "", "output": "私はローカルAIアシスタントです。"}
```

### 保存例：

```
./data/train.jsonl
```

---

## ✅ 2. モデルとトークナイザの読み込み

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import get_peft_model, LoraConfig, TaskType

base_model = "Qwen/Qwen3-4B-Instruct"

tokenizer = AutoTokenizer.from_pretrained(base_model, trust_remote_code=True)
model = AutoModelForCausalLM.from_pretrained(
    base_model, load_in_8bit=True, device_map="auto", trust_remote_code=True
)
```

---

## ✅ 3. LoRAの設定

```python
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["c_attn", "mlp.gate_proj"],  # モデル構造に応じて調整
    lora_dropout=0.05,
    bias="none",
    task_type=TaskType.CAUSAL_LM
)

model = get_peft_model(model, lora_config)
```

---

## ✅ 4. データローダとTrainer構築

```python
from datasets import load_dataset
from transformers import TrainingArguments, Trainer

# JSONLの読み込み
dataset = load_dataset("json", data_files={"train": "./data/train.jsonl"})["train"]

# トークナイズ
def preprocess(example):
    prompt = example["instruction"]
    if example["input"]:
        prompt += "\n" + example["input"]
    prompt += "\n" + example["output"]
    tokens = tokenizer(prompt, truncation=True, padding="max_length", max_length=1024)
    tokens["labels"] = tokens["input_ids"].copy()
    return tokens

tokenized = dataset.map(preprocess, remove_columns=dataset.column_names)

training_args = TrainingArguments(
    output_dir="./lora-qwen3",
    per_device_train_batch_size=2,
    num_train_epochs=3,
    logging_steps=10,
    save_strategy="epoch",
    learning_rate=2e-4,
    bf16=True,
    optim="paged_adamw_8bit"
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized
)

trainer.train()
```

---

## ✅ 5. LoRAモデルの保存

```python
model.save_pretrained("./lora-qwen3")
tokenizer.save_pretrained("./lora-qwen3")
```

---

## ✅ 6. 推論時の使い方（LoRA適用）

```python
from peft import PeftModel
from transformers import AutoTokenizer, AutoModelForCausalLM

base = AutoModelForCausalLM.from_pretrained(base_model, device_map="auto", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained(base_model, trust_remote_code=True)
lora = PeftModel.from_pretrained(base, "./lora-qwen3")
lora.eval()

inputs = tokenizer("君の名前は？", return_tensors="pt").to("cuda")
outputs = lora.generate(**inputs, max_new_tokens=50)
print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

---

## ✅ 7. Ollama用に変換（任意）

1. `transformers-to-gguf` を使って GGUF形式に変換
2. `Modelfile` を作って `ollama create` で登録

```bash
transformers-to-gguf --model ./lora-qwen3 --output ./qwen3-lora.gguf --dtype q4_0
```

---

## ✅ ライセンス

この手順で使用するQwen3は Apache 2.0 ライセンスで、商用利用が可能です。  
必ず元モデルの [ライセンス条件](https://huggingface.co/Qwen/Qwen3-4B-Instruct) をご確認ください。

---

## ✅ 補足情報

- パラメータ量が多いため、学習には最低 VRAM 8GB（可能なら12GB以上）推奨です。
- `target_modules` の指定はモデル構造によって変わるため、`model.named_modules()` で確認をおすすめします。
- `peft==0.10.0` 以降の安定バージョン推奨です。

---
```

---

必要に応じて、このMarkdownをファイル（`README.md`）形式でダウンロード可能にもできます。ご希望ですか？
